### PR TITLE
[release/2.2] Fix ReadAsync not returning when cancellation token fires during FlushAsync

### LINKED ISSUE #31201: ReadAsync might never return if FlushAsync cancellation token is cancelled
It has to happen while FlushAsync is running.

### PR DESCRIPTION
**Pull request by pakrym**
Fixes #31201

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs [hunk 1]
```diff
@@ -60,14 +60,23 @@ public CancellationTokenRegistration AttachToken(CancellationToken cancellationT
...
         {
             Action<object> currentCompletion = _completion;
+            object currentState = _completionState;
+
             _completion = s_awaitableIsCompleted;
+            _completionState = null;
 
             completionData = default;
-
+
```
Comment (davidfowl) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs]: nit: whitespace

**Comment by karelz** [comment]
@danmosemsft the PR is against 2.2, yet it has approved-2.1.5 label. What am I missing?
Do we need to manually port fixes into 2.2?

**Comment by danmosemsft** [comment]
@karelz good point I did not notice that. @pakrym this was approved for 2.1.5. Please create a PR against release/2.1 and then move this servicing-approved-2.1.5 label to that PR.

Submissions into release/2.1 do flow to release/2.2 however with the branch not open for a while if you need it there sooner it makes sense to merge this PR also.

**Comment by karelz** [comment]
> however with the branch not open for a while if you need it there sooner it makes sense to merge this PR also

@pakrym I would discourage from merging it into 2.2 branch, unless it is really, really needed there ASAP (in which case, please provide business justification). The key reason is to keep things simple and to keep integration costs lower further on. Thanks!

**Comment by danmosemsft** [comment]
@pakrym can you please link the 2.1 PR? I don't see it.

**Comment by danmosemsft** [comment]
You can mark it servicing-approved-2.1.5 if it is ready fairly soon.

**Pull request closed without merging**